### PR TITLE
Fix miscellaneous Clang and Cppcheck 2.9 style warnings

### PR DESCRIPTION
**Pull request by seanm**
I'm really not familiar with the codebase, so please review carefully.  libexpat is a dependency of a dependency of a dependency for me, and I'm just trying to get everything building with more warnings fixed.

The `_GNU_SOURCE` thing in particular could maybe be addressed differently.  Perhaps it doesn't need to be defined at all?

### CODE REVIEW

**Review by hartwork — CHANGES_REQUESTED** [review]
Hi @seanm,


I have reviewed all but the `-Wcast-qual` fixes now, please see below for my findings.

Given the volume of both fixed (and unfixed) warnings regarding `-Wcast-qual` I'm considering to suggest dropping or extracting the two related commits because of the unhealthy value-per-time aspect of those changes. It would allow us to finish and merge the rest of the pull request quicker.

Other than `-Wcast-qual`, here is what I found during review:

- Commit `Fixed all clang -Wconditional-uninitialized warnings`
  - No concerns or objections.
- Commit `Fixed most clang -Wreserved-id-macro warnings`
  - It's apparently all not just most; if true please adjust the commit message
  - No other concerns or objections.
- Commit `Fixed all clang -Wunused-macros warnings`
  - `poolEnd` — confirmed unused
  - `SET2` — confirmed unused since commit 583f937102f5f95e41dbd0e6f4ffb9d3323db3aa.
    Found using `git log -p | grep -E 'SET2|^commit ' | grep -B1 SET2`
  - `XML_FMT_CHAR` — confirmed unused
  - :warning:  `DUMMY_DEFAULT_HANDLER_FLAG` — technically unused, may need a line `dummy_handler_flags |= DUMMY_DEFAULT_HANDLER_FLAG` at the end of function `dummy_default_handler`.  Three more functions need a closer look (or a maybe comment) for the same thing:

    - `dummy_xdecl_handler`
    - `dummy_end_element`
    - `dummy_cdata_handler`

    CC @RMJ10 
- Commit `Fixed all clang -Wunreachable-code warnings`
  - :warning: Those `0 && ` were added on purpose (see 5841ab111b4cd0528aadc4e9c93d675fde39aa3f) so that the disabled code is still compiled. So that's a feature so this commit should be dropped to not make things worse while we still have that code.
- Commit `Fixed some clang -Wcast-qual warnings`
  - TODO/ignored as mentioned above
- Commit `Fixed a clang -Wcast-qual warning with strtoul() usage`
  - TODO/ignored as mentioned above
- Commit `Fixed cppcheck warnings`:
  - My local Cppcheck 2.9 does not produces these warnings.  Please elaborate on which version of Cppcheck you are using and how you are invoking the tool so I can see these warnings myself.  Thank you!
  - :warning: There is an unrelated change from `sprintf` to `snprintf` in this commit.  Please get the unrelated change out of this commit (e.g using `git rebase -i`). Thank you!
  - (Let me do more research on this:) My vote against pulling `int i = 0` into the for loop header.  Let's do `int i = 0` right before the `for` loop and keep the first `for` header segment empty a la `for (; update; check)`. What do you think?
- Commit `Replaced all sprintf with safer snprintf`
  - I have a phantom memory that use of `snprintf` causes portability issues due to difference in prototype across platforms.  I cannot find or confirm right now, but I will need to check again.
- Commit `Fixed a clang -Wreserved-id-macro warnings by conditionalizing to linux-only`
  - :warning: To my understanding `_GNU_SOURCE` is about glibc and not about Linux and limiting the scope is not a proper fix to the warning in my eyes. Please drop this commit. If we want to improve on it, we can, but it deserves a dedicated issue and or pull request then.

Best, Sebastian

**Comment by seanm** [comment]
OK, thanks for your comprehensive review and comments... gimme a couple of hours and I'll have a reply and new commits pushed...

**Comment by hartwork** [comment]
@seanm three more notes on `snprintf`.

First, while `snprintf`'s return value may not be too relevant to our case in Expat, `snprintf` returning 5 here (on Linux) scares me:

```console

# gcc -std=c99 -Wextra -pedantic main.c && ./a.out
snprintf returned: 5 !!!
[0] 0x68
[1] 0x65
[2] 0x00

# cat main.c 
#include <assert.h>
#include <stdio.h>

int main() {
    char buf[3];
    const int bytesWritten = snprintf(buf, sizeof(buf), "hello");
    printf("snprintf returned: %d !!!\n", bytesWritten);
    assert(bytesWritten >= 0);
    int i = 0;
    for (; i < (int)sizeof(buf); i++) {
        printf("[%d] 0x%02x\n", i, buf[i]);
    }
    return 0;
}
```

And this part of the MSDN docs at https://learn.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2012/2ts7cx93(v=vs.110)?redirectedfrom=MSDN seems worth noting, emphasis mine:

> If len = count, then len characters are stored in buffer, **no null-terminator is appended, and len is returned**.
> 
> If len > count, then count characters are stored in buffer, **no null-terminator is appended, and a negative value is returned**.

I cannot it conform it easily, but it doesn't exactly make `snprintf` a likely candidate for a friend of mine.

I'm tempted to say that the warning is the problem rather than the current code: because Expat's use of `snprintf` seems safe and sane and general use (and advertising) of `snprintf` by the compiler seems questional to me, given the fidings above.  Maybe that's why it's not included with `-Wextra`.

**Comment by hartwork** [comment]
@seanm update on the `for` loop variable declaration: I can now confirm that `for (int i = 3; i < 5; i++)` is legal in >=C99 but not earlier. Expat request >=C99 officially by now, so technically that code is an option to us, for better or worse. For proof:

```console
# gcc -std=c89 -Wextra -pedantic main.c && ./a.out
main.c: In function ‘main’:
main.c:4:5: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
    4 |     for (int i = 3; i < 5; i++) {
      |     ^~~
main.c:4:5: note: use option ‘-std=c99’, ‘-std=gnu99’, ‘-std=c11’ or ‘-std=gnu11’ to compile your code

# cat main.c 
#include <stdio.h>

int main() {
    for (int i = 3; i < 5; i++) {
        printf("i is %d\n", i);
    }
    return 0;
}
```

Related links that I consider an interesting read:
- [What are the C++ scoping rules for for loops?](https://www.cs.technion.ac.il/users/yechiel/c++-faq/scope-of-for-loop-var.html)
- [Moving the kernel to modern C](https://lwn.net/Articles/885941/)

**Comment by seanm** [comment]
>Commit Fixed most clang -Wreserved-id-macro warnings
>It's apparently all not just most; if true please adjust the commit message

In fact it is just most, because it doesn't fix `_GNU_SOURCE`, which as you saw is in another commit because the fix for it is dubious I agree.

I've dropped `Fixed all clang -Wunreachable-code warnings` as requested.

I'll make a new PR for the `consts`.

> Please elaborate on which version of Cppcheck you are using and how you are invoking the tool so I can see these warnings myself

Version 2.9: `cppcheck --project=compile_commands.json --platform=unspecified --enable=style -q --library=qt --library=posix --library=gnu --library=bsd --library=windows --template='{id},{file}:{line},{severity},{message}' -j8`

>My vote against pulling int i = 0 into the for loop header. Let's do int i = 0 right before the for loop and keep the first for header segment empty a la for (; update; check). What do you think?

I disagree.  `for (int i = 0; i < count; i++)` is idiomatic and limits the scope of `i` to inside the loop.  Do you need to support pre-C99?  It a [common coding guideline](https://wiki.sei.cmu.edu/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions).

>First, while snprintf's return value may not be too relevant to our case in Expat, snprintf returning 5 here (on Linux) scares me

It's documented to: "return the number of characters that ***would have*** been printed if the size were unlimited". This lets you know the buffer you supplied was not big enough.  Otherwise how would you know your string got truncated? Anyway, the return value both before and after my change was never even looked at.

>If len = count, then len characters are stored in buffer, no null-terminator is appended, and len is returned.

That's `_snprintf` (note the leading underscore). That's an older function from before MS supported the standard `snprintf`.  The standard `snprintf` is never worse than the standard `sprintf`.  Except for the return value (which is often not used) the only difference if that `snprintf` will never overrun your buffer because you tell it how big the buffer is.  It also ***always*** null terminates, truncating if necessary.

>Commit Fixed a clang -Wreserved-id-macro warnings by conditionalizing to linux-only 

I'll create a ticket for this.

**Comment by seanm** [comment]
For `_GNU_SOURCE`, I'm thinking maybe:

```
#if defined(HAVE_SYSCALL_GETRANDOM)
#  define _GNU_SOURCE 1 /* syscall prototype */
#endif
```

Since it's being defined only for `syscall()`.

**Comment by hartwork** [comment]
> For `_GNU_SOURCE`, I'm thinking maybe:
> 
> ```
> #if defined(HAVE_SYSCALL_GETRANDOM)
> #  define _GNU_SOURCE 1 /* syscall prototype */
> #endif
> ```
> 
> Since it's being defined only for `syscall()`.

@seanm we *can* add that, provided we keep the already-defined guard in place. I'd wonder though what practical issue we'd fix by that. We would not fully fix the warning, except maybe for where `HAVE_SYSCALL_GETRANDOM` is not defined. Id say we either fit it for all platforms or none.

**Comment by hartwork** [comment]
> > Commit Fixed most clang -Wreserved-id-macro warnings
> > It's apparently all not just most; if true please adjust the commit message
> 
> In fact it is just most, because it doesn't fix `_GNU_SOURCE`, which as you saw is in another commit because the fix for it is dubious I agree.

@seanm understood, you're right.

> I've dropped `Fixed all clang -Wunreachable-code warnings` as requested.

Thanks!

> I'll make a new PR for the `consts`.

Thanks!


> > Please elaborate on which version of Cppcheck you are using and how you are invoking the tool so I can see these warnings myself
> 
> Version 2.9: `cppcheck --project=compile_commands.json --platform=unspecified --enable=style -q --library=qt --library=posix --library=gnu --library=bsd --library=windows --template='{id},{file}:{line},{severity},{message}' -j8`

Seems like it's `--enable=style` that enabled those. We can include those that you already have, but let's refrain from puilling in more like theses because the value these fixes bring doesn't make up for the time you and me will have put in by the end of it.

> > My vote against pulling int i = 0 into the for loop header. Let's do int i = 0 right before the for loop and keep the first for header segment empty a la for (; update; check). What do you think?
> 
> I disagree. `for (int i = 0; i < count; i++)` is idiomatic and limits the scope of `i` to inside the loop. Do you need to support pre-C99? It a [common coding guideline](https://wiki.sei.cmu.edu/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions).

On the C99 part please see https://github.com/libexpat/libexpat/pull/669#issuecomment-1292361623 that I posted in the mean time. I hope that `-Wextra` would wake us up if we ever run into variable shadowing from that technique. Your call.

> > First, while snprintf's return value may not be too relevant to our case in Expat, snprintf returning 5 here (on Linux) scares me
> 
> It's documented to: "return the number of characters that _**would have**_ been printed if the size were unlimited". This lets you know the buffer you supplied was not big enough. Otherwise how would you know your string got truncated?

That's an interesting point. So unfortunate but necessary, I suppose. (Honestly, I still favor a proper check and use of `sprintf`. The code will be safe and explain itself.)

> Anyway, the return value both before and after my change was never even looked at.

Yes. (I'm acknowledging that at https://github.com/libexpat/libexpat/pull/669#issuecomment-1292337052 that I posted in the mean time.)

> > If len = count, then len characters are stored in buffer, no null-terminator is appended, and len is returned.
> 
> That's `_snprintf` (note the leading underscore). That's an older function from before MS supported the standard `snprintf`. The standard `snprintf` is never worse than the standard `sprintf`. Except for the return value (which is often not used) the only difference if that `snprintf` will never overrun your buffer because you tell it how big the buffer is. It also _**always**_ null terminates, truncating if necessary.

Interesting point, I think you're right. So https://learn.microsoft.com/en-us/previous-versions/2ts7cx93(v=vs.140) would have been a better link for me to share and study.

> > Commit Fixed a clang -Wreserved-id-macro warnings by conditionalizing to linux-only
> 
> I'll create a ticket for this.

Okay.

**Comment by hartwork** [comment]
@seanm regarding feature tests for `snprintf` I find…

> ```
>        snprintf(), vsnprintf():
>            _XOPEN_SOURCE >= 500 || _ISOC99_SOURCE
>                || /* Glibc <= 2.19: */ _BSD_SOURCE
> ```

…in `man 3 snprintf` and …

> Macro: _ISOC99_SOURCE
>
> If this macro is defined, features from ISO C99 are included. Since these features are included by default, this macro is mostly relevant when the compiler uses an earlier language version. 

at https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html . I guess we don't need to enforce `_ISOC99_SOURCE` ourselves then. What do you think?

**Comment by seanm** [comment]
>Seems like it's --enable=style that enabled those. We can include those that you already have, but let's refrain from puilling in more like theses

There are no more.  cppcheck gives no warnings after this PR.

>(Honestly, I still favor a proper check and use of sprintf. The code will be safe and explain itself.)

There is no such thing as safe use of `sprintf`. :)  It's just too easy to overrun a buffer.  `sprintf` should ***never*** be used.

>I guess we don't need to enforce _ISOC99_SOURCE ourselves then. What do you think?

I don't think you need `_ISOC99_SOURCE`.  C99 requires `snprintf`, you require C99. Done. :)

>we can add that, provided we keep the already-defined guard in place.

Done.  I've brought it back in this PR.

>I'd wonder though what practical issue we'd fix by that. We would not fully fix the warning, except maybe for where HAVE_SYSCALL_GETRANDOM is not defined.

Which is many places: macOS, Windows, BSD, etc.  That's valuable to me.  It means I have no more -Wreserved-id-macro warnings on macOS anymore.

>Id say we either fit it for all platforms or none.

It's impossible to fix for linux/glibc, since they require it to be defined.

File: expat/tests/runtests.c [hunk 1]
```diff
@@ -199,7 +197,6 @@ static unsigned long dummy_handler_flags = 0;
 #define DUMMY_START_DOCTYPE_DECL_HANDLER_FLAG (1UL << 14)
 #define DUMMY_END_DOCTYPE_DECL_HANDLER_FLAG (1UL << 15)
 #define DUMMY_SKIP_HANDLER_FLAG (1UL << 16)
-#define DUMMY_DEFAULT_HANDLER_FLAG (1UL << 17)
```
[the remark this answers is not in the document]
Comment (seanm) [reply expat/tests/runtests.c:202]: Done.

[the remark this answers is not in the document]
Comment (hartwork) [reply expat/tests/runtests.c]: Thank you!

**Comment by hartwork** [comment]
> > Seems like it's --enable=style that enabled those. We can include those that you already have, but let's refrain from puilling in more like theses
> 
> There are no more. cppcheck gives no warnings after this PR.

Cannot confirm. When I do `git ls-files \*.c \*.cpp \*.h | xargs cppcheck --enable=style` on this code I get more warnings e.g. including `The scope of the variable '...' can be reduced.` Maybe it's a different  in config, but let's not pull more of these in either way.

> > (Honestly, I still favor a proper check and use of sprintf. The code will be safe and explain itself.)
> 
> There is no such thing as safe use of `sprintf`. :) It's just too easy to overrun a buffer. `sprintf` should _**never**_ be used.

We don't agree on that then.

> > I guess we don't need to enforce _ISOC99_SOURCE ourselves then. What do you think?
> 
> I don't think you need `_ISOC99_SOURCE`. C99 requires `snprintf`, you require C99. Done. :)

Alright :smile: 

> > we can add that, provided we keep the already-defined guard in place.
> 
> Done. I've brought it back in this PR.
> 
> > I'd wonder though what practical issue we'd fix by that. We would not fully fix the warning, except maybe for where HAVE_SYSCALL_GETRANDOM is not defined.
> 
> Which is many places: macOS, Windows, BSD, etc. That's valuable to me. It means I have no more -Wreserved-id-macro warnings on macOS anymore.
> 
> > Id say we either fit it for all platforms or none.
> 
> It's impossible to fix for linux/glibc, since they require it to be defined.

I feel a bit like my argument has been ignored here. Fixing some warnings upstream is a good idea but compiling dependencies with the same warning level as your first party code is not something I consider useful, these two builds schould be isolated. The only reason I'm not asking for a revert of that addition is because a correctness argument can be made that is hard to argue against. It is however by no means something worth calling a fix to the warning, it only hides the code from analysis.

**Comment by seanm** [comment]
>Cannot confirm. When I do git ls-files \*.c \*.cpp \*.h | xargs cppcheck --enable=style on this code I get more warnings

Hmmm, strange.  We are invoking it differently though. I'm using CMake's `CMAKE_EXPORT_COMPILE_COMMANDS=ON` to generate `compile_commands.json` and then passing `--project=compile_commands.json` to cppcheck.  I guess that's somehow the difference...

>We don't agree on that then.

*If the return value is unneeded (like in libexpat)*, when is `sprintf` preferable to `snprintf`?  I honestly cannot think of any situation.

**Comment by hartwork** [comment]
>>> There is no such thing as safe use of `sprintf`. :)  It's just too easy to overrun a buffer.  `sprintf` should ***never*** be used.

>>    We don't agree on that then.

> [..], when is `sprintf` preferable to `snprintf`? I honestly cannot think of any situation.

@seanm you said above there is no safe way to use `sprintf`… but all uses of `sprintf` we just patched in this pull request were perfectly safe. "Preferable" might by arguable or taste, but there was no issue to be fixed here, any warning about it had no point beyond _general awareness_ on these very cases.

**Comment by RMJ10** [comment]
>       * warning  `DUMMY_DEFAULT_HANDLER_FLAG` — technically unused, may need a line `dummy_handler_flags |= DUMMY_DEFAULT_HANDLER_FLAG` at the end of function `dummy_default_handler`.  Three more functions need a closer look (or a maybe comment) for the same thing:
>         
>         * `dummy_xdecl_handler`
>         * `dummy_end_element`
>         * `dummy_cdata_handler`

I think `DUMMY_DEFAULT_HANDLER_FLAG` may have got added on the assumption that it was needed for a test.

The purpose of the dummy handlers, as the comment says, is to provide a handler when we need one but don't care what it does.  Setting flags to prove the handler has been run is extra to that, so not having a flag set by those three functions isn't wrong.  It's safe to delete the definition of `DUMMY_DEFAULT_HANDLER_FLAG` and for now it's probably the right answer.

Strictly speaking we should probably have a bunch of really basic tests that check the handlers do fire when they're supposed to, but so many other tests would fail if they didn't that it's not really worth while to do that.

**Comment by hartwork** [comment]
@RMJ10 thanks for jumping in and the elaboration. Summary note to myself: It's sound to drop the `DUMMY_DEFAULT_HANDLER_FLAG` macro.

**Comment by seanm** [comment]
>you said above there is no safe way to use sprintf…

OK, I guess that way maybe a little too strong. I suppose use of `sprintf` can sometimes have no bugs.

>but all uses of sprintf we just patched in this pull request were perfectly safe, already.

Honestly, I didn't even check.  But checking now, it was easy to find at least one was buggy:

```
static void
_xml_failure(XML_Parser parser, const char *file, int line) {
  char buffer[1024];
  enum XML_Error err = XML_GetErrorCode(parser);
  sprintf(buffer,
          "    %d: %" XML_FMT_STR " (line %" XML_FMT_INT_MOD
          "u, offset %" XML_FMT_INT_MOD "u)\n    reported from %s, line %d\n",
          err, XML_ErrorString(err), XML_GetCurrentLineNumber(parser),
          XML_GetCurrentColumnNumber(parser), file, line);
  _fail_unless(0, file, line, buffer);
}
```

The `file` path parameter could easily be more than ~900 characters, and overflow the 1024 byte buffer.

>"Preferable" might by arguable or taste, but there was no issue to be fixed here, any warning about it had no point beyond general awareness, on these very cases.

"Preferable" is too weak. I maintain that there is never a good reason to choose `sprintf` instead of `snprintf`.

**Comment by hartwork** [comment]
> > you said above there is no safe way to use sprintf…
> 
> OK, I guess that way maybe a little too strong. I suppose use of `sprintf` can sometimes have no bugs.

@seanm thanks for saying that.

> > but all uses of sprintf we just patched in this pull request were perfectly safe, already.
> 
> Honestly, I didn't even check. But checking now, it was easy to find at least one was buggy:
> 
> ```
> static void
> _xml_failure(XML_Parser parser, const char *file, int line) {
>   char buffer[1024];
>   enum XML_Error err = XML_GetErrorCode(parser);
>   sprintf(buffer,
>           "    %d: %" XML_FMT_STR " (line %" XML_FMT_INT_MOD
>           "u, offset %" XML_FMT_INT_MOD "u)\n    reported from %s, line %d\n",
>           err, XML_ErrorString(err), XML_GetCurrentLineNumber(parser),
>           XML_GetCurrentColumnNumber(parser), file, line);
>   _fail_unless(0, file, line, buffer);
> }
> ```
> 
> The `file` path parameter could easily be more than ~900 characters, and overflow the 1024 byte buffer.

There is potentially no such case in the code base given that all calls to `_xml_failure` come from hard-coded tests and use the value of `__FILE__` most or all of the time, but you have a point nonetheless.

**Pull request merged**